lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs

```diff
@@ -0,0 +1,488 @@
+
+    -- * Deconstruction
+    , toList
+    , toUTxO
+
+    -- * Folding
+    , fold
+
+    -- * Modification
+    , insert
+    , insertMany
+    , delete
+    , deleteMany
+
+    -- * Queries
+    , assets
+    , balance
+    , lookup
+    , member
+    , null
+    , size
+
+    -- * Selection
+    , SelectionFilter (..)
+    , selectRandom
+
+    ----------------------------------------------------------------------------
+    -- Internal Interface
+    ----------------------------------------------------------------------------
+
+    -- * Utilities
+    , selectRandomSetMember
+
+    -- * Invariant
+    , InvariantStatus (..)
+    , checkInvariant
+
+    ) where
+
+import Prelude hiding
+    ( lookup, null )
+
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle )
+import Cardano.Wallet.Primitive.Types.TokenMap
+    ( AssetId )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( TxIn, TxOut )
+import Cardano.Wallet.Primitive.Types.UTxO
+    ( UTxO (..) )
+import Control.DeepSeq
+    ( NFData )
+import Control.Monad.Random.Class
+    ( MonadRandom (..) )
+import Data.Generics.Internal.VL.Lens
+    ( view )
+import Data.Generics.Labels
+    ()
+import Data.Map.Strict
+    ( Map )
+import Data.Maybe
+    ( isJust )
+import Data.Set
+    ( Set )
+import Data.Set.Strict.NonEmptySet
+    ( NonEmptySet )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Data.Foldable as F
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import qualified Data.Set.Strict.NonEmptySet as NonEmptySet
+
+--------------------------------------------------------------------------------
+-- Public Interface
+--------------------------------------------------------------------------------
+
+--------------------------------------------------------------------------------
+-- Type
+--------------------------------------------------------------------------------
+
+-- | A UTxO set that is indexed by asset identifier.
+--
+-- The index provides a mapping from assets to sets of transaction inputs.
+-- A transaction input appears in the set if and only if its corresponding
+-- output contains a non-zero quantity of that asset.
+--
+-- The index makes it possible to efficiently compute the subset of a UTxO
+-- set containing a particular asset, or to select a UTxO entry containing
+-- that asset, without having to search through the entire UTxO set.
+--
+-- The index also keeps track of the current UTxO balance for all assets,
+-- making it possible to efficiently look up the quantity of a particular
+-- asset without having to sum across the entire UTxO set.
+--
+-- The UTxO index data structure has an invariant that can be checked with
+-- the 'checkInvariant' function.
+--
+data UTxOIndex = UTxOIndex
+    { index
+        :: !(Map AssetId (NonEmptySet TxIn))
+    , balance
+        :: !TokenBundle
+    , utxo
+        :: !(Map TxIn TxOut)
```

Comment:
> Why is this not using newtype UTxO = UTxO { getUTxO :: Map TxIn TxOut }?

We could, but it would add a layer of wrapping and unwrapping to every operation that manipulates the `utxo` field.

The current implementation only needs to do this wrapping and unwrapping just once, if we actually need to convert to from an ordinary `UTxO`:
```
toUTxO :: UTxOIndex -> UTxO
toUTxO = UTxO . utxo

fromUTxO :: UTxO -> UTxOIndex
fromUTxO = Map.foldlWithKey' (\u i o -> insertUnsafe i o u) empty . getUTxO
```